### PR TITLE
add NO_REPEAT_RANDOM option to SelectFloat()

### PR DESCRIPTION
**Pull request by fdarosa2663**
Open to improvements but this implementation is working to randomize the up/down selection if SelectFloat() tries to repeat file as a start point.

### CODE REVIEW

File: sound/effect.h [hunk 1]
```diff
@@ -231,9 +231,31 @@ class Effect {
     selected_++;
     if (selected_ == (int)files_found()) selected_ = 0;
   }
+#ifdef NO_REPEAT_RANDOM
+  int last_selected_;
```
Comment (profezzorn) [inline sound/effect.h:235]: No, we are not going to add a new variable, we're going to fix how last_ works instead.

File: sound/effect.h [hunk 2]
```diff
@@ -231,9 +231,31 @@ class Effect {
...
+#endif
   void SelectFloat(float value) {
     int f = files_found();
     selected_ = clamp(floorf(f * value), 0, f - 1);
+#ifdef NO_REPEAT_RANDOM
+    if (selected_ == last_selected_) {
+      if (f == 2) {
+        selected_ = rand() % num_files_;
```
Comment (profezzorn) [inline sound/effect.h:243]: just se selected_ to -1 and let RandomFile do the picking.

File: sound/effect.h [hunk 3]
```diff
@@ -231,9 +231,31 @@ class Effect {
...
+      } else {
+        if (selected_ == (int)files_found()) {
```
Comment (profezzorn) [inline sound/effect.h:245]: This is never going to be true, since selected_ is in the range 0 ... f - 1

File: sound/effect.h [hunk 4]
```diff
@@ -231,9 +231,31 @@ class Effect {
...
+          selected_ -= 1;
```
Comment (profezzorn) [inline sound/effect.h:246]: We could get in trouble here if there is only one file.

File: sound/effect.h [hunk 5]
```diff
@@ -231,9 +231,31 @@ class Effect {
...
+        } else {
+          int n = rand() % 2;
+          if (selected_ >= 1 && n == 2) {
```
Comment (profezzorn) [inline sound/effect.h:249]: rand() % 2 will never be equal to 2

File: sound/effect.h [hunk 6]
```diff
@@ -231,9 +231,31 @@ class Effect {
...
     selected_ = clamp(floorf(f * value), 0, f - 1);
+#ifdef NO_REPEAT_RANDOM
+    if (selected_ == last_selected_) {
+      if (f == 2) {
+        selected_ = rand() % num_files_;
+      } else {
+        if (selected_ == (int)files_found()) {
+          selected_ -= 1;
+        } else {
+          int n = rand() % 2;
```
Comment (profezzorn) [inline sound/effect.h:248]: no need for a variable here, just put rand() % 2 in the next expression.

File: sound/effect.h (hunk 1 above)
Comment (fdarosa2663) [reply sound/effect.h:235]: no problem, will need last_ fixed as it rewrites between Select currently.

File: sound/effect.h (hunk 2 above)
Comment (fdarosa2663) [reply sound/effect.h]: done

File: sound/effect.h (hunk 3 above)
Comment (fdarosa2663) [reply sound/effect.h]: my mistake, fixed

File: sound/effect.h (hunk 4 above)
Comment (fdarosa2663) [reply sound/effect.h:243]: added "f > 1" to upper if statement

File: sound/effect.h (hunk 5 above)
Comment (fdarosa2663) [reply sound/effect.h]: fixed

File: sound/effect.h (hunk 6 above)
Comment (fdarosa2663) [reply sound/effect.h]: fixed

File: sound/effect.h (hunk 1 above)
Comment (profezzorn) [reply sound/effect.h]: so fix it?

**Comment by profezzorn** [comment]
I implemented this myself a bit differently.

**Pull request closed without merging**